unix: Include upip as fronzen modules inside the standard interpreter.

pfalcon opened the pull request:
MicroPython doesn't come with standard library included, so it is important
to be able to easily install needed package in a seamless manner. Bundling
package manager (upip) inside an executable solves this issue.

upip is bundled only with standard executable, not "minimal" or "fast"
builds.

Commit:
unix: Include upip as fronzen modules inside the standard interpreter.

MicroPython doesn't come with standard library included, so it is important
to be able to easily install needed package in a seamless manner. Bundling
package manager (upip) inside an executable solves this issue.

upip is bundled only with standard executable, not "minimal" or "fast"
builds.

dpgeorge commented:
> MicroPython doesn't come with standard library included

Maybe it should?  Or at least the unix version (which is what this discussion is all about).  You say that "Python is batteries included", and I agree, and so maybe uPy should be "micro-batteries included".  If we want proper Python compatibility then the most obvious way is to have the stdlib as part of the standard distribution.  Well, you should still be able to run uPy without any support file (like it is now), but then the next step if you want more features is to install all of micropython-lib by getting and unpacking a single file.  Then upip is left for the non-standard micro-libraries, as per pip.  This makes uPy the same as CPy, and seems to me to be the most obvious thing to do.

It seems a bit of a burden to install each stdlib separately using upip (yes you can use requirements.txt, but then why not just list all libs in requirements.txt and be done with it?).  When running on a PC you most likely have a few MB spare to store all the library scripts, and having them there does not slow down uPy or make it use more RAM if you never use the libs.

dpgeorge commented on unix/Makefile:
```diff
@@ -153,3 +158,9 @@ coverage_test: coverage
+	../tools/make-frozen.py frozen_upip > $@
+
+frozen_upip/upip.py:
+	../tools/bootstrap_upip.sh frozen_upip
```
This is no good, since now the standard unix build requires internet connectivity (and wget).  I'm strongly -1 on that.  See eg Travis failure.

pfalcon replied:
Sure, I noted myself too that it will complicate my work on a beach/plane. I'm happy to duplicate bytes and drop upip tarball (easiest and "most contained" solution) or upip sources into codebase - if you're happy with such duplication.

pfalcon commented:
> Maybe it should? Or at least the unix version (which is what this discussion is all about). You say that "Python is batteries included", and I agree, and so maybe uPy should be "micro-batteries included".

But how it's different from CPython then? Installation size will be dominated by library size then, and it will be the same bloat as CPython. The whole idea is to stay "micro", while being Python (i.e. being able to do all the same what CPYthon does).

> but then the next step if you want more features is to install all of micropython-lib

Sure, if someone can afford that, they will run "micropython -m upip install micropython-lib-all" as the next step after getting an executable (when micropython-lib-all package is contributed by those who need it).

> When running on a PC you

When running on a PC you don't care about MicroPython at all. We should not cater for people who will never care to use uPy, we should cater for people who may be interested by its minimalism.

dpgeorge commented:
I agree that there's something satisfying about having the micropython executable self contained to the point where it can install its own libraries.  So I agree to merging this.

But we should definitely put the upip tarball in this repo to avoid need for internet connectivity for a build.  To that end, upip should be minimised as much as practical (I see that at the moment the upip_XX modules are just cut-and-paste from the stdlib; they can be reduced).

Also note that this won't work on the windows build (for many reasons) and that's something to tackle one day.

dpgeorge commented on unix/Makefile:
```diff
@@ -103,6 +103,11 @@ SRC_C = \
 	coverage.c \
 	$(SRC_MOD)
 
+# Include builtin package manager is the standard build
```
"is" -> "in.

pfalcon replied:
Fixed.

Commit:
tools: Add upip 0.5.2 tarball.

So unix version, which now includes upip as a frozen module, can be built
without Internet connection.

Commit:
unix: Use upip tarball in tools/ to make upip frozen module.

Instead of downloading it from PyPI, which requires Internet connection
which may not be available in all build environments.

pfalcon commented:
> But we should definitely put the upip tarball in this repo to avoid need for internet connectivity for a build.

Done. So, update workflow will be: 1) develop upip in micropython-lib; 2) use makefile there to pull any updates to dependency modules; 3) release to PyPI; 4) download tarball and commit to micropython repo.

pfalcon commented:
>  To that end, upip should be minimised as much as practical

Yes, I keep that in mind, that's for v2 of this feature. (upip needs further love either, like error reporting, etc.)

pfalcon commented:
> Also note that this won't work on the windows build (for many reasons) and that's something to tackle one day.

Yes, and as we can't rely on wget on Windows, dealing with SSL is on critical path to resolving that.

pfalcon commented:
So, patch is updated, please have a look.

Commit:
unix: Coverage build should also include upip as frozen module.

Purely for consistency, coverage build was the same as the standard, with
coverage enabled is the only difference, so after adding frozen module it
should stay the same.

pfalcon commented:
Build fixed, by applying a bit hacky fix for coverage build to also include frozen upip.

dpgeorge commented:
> by applying a bit hacky fix for coverage build to also include frozen upip.

I don't think it's hacky.

dpgeorge commented on unix/Makefile:
```diff
@@ -153,3 +160,16 @@ coverage_test: coverage
+	tar xfz $^
+	rm -rf frozen_upip
+	mkdir -p frozen_upip
+	cp micropython-upip-*/upip*.py frozen_upip/
```
Perhaps prefix these commands with $(Q) and print a nice message at the start like $(ECHO) "UNPACK upip".

pfalcon replied:
Done.

dpgeorge commented:
Please add micropython-upip-* to .gitignore.

I assume you'll squash the commits into one.

I think we can get rid of the old pip stuff in the makefile (install and uninstall of tools/pip-micropython).

dpgeorge commented on unix/Makefile:
```diff
@@ -153,3 +160,16 @@ coverage_test: coverage
 	gcov -o build-coverage/py ../py/*.c
 	gcov -o build-coverage/extmod ../extmod/*.c
+
+_frozen_upip.c: frozen_upip/upip.py
```
Can _frozen_upip.c go in the $(BUILD)/ directory?  Would be cleaner that way (since it's a generated file).

pfalcon replied:
At the expense of .o ending up in $(BUILD)/$(BUILD)/ , possibly it could.

pfalcon replied:
Done

pfalcon commented:
> Please add micropython-upip-* to .gitignore.

Done

pfalcon commented:
> I think we can get rid of the old pip stuff in the makefile (install and uninstall of tools/pip-micropython).

Let's serialize that a bit I guess.

Commit:
frozen upip: Latest changes per review (to be squashed).

pfalcon commented:
Squashed and merged.

dpgeorge commented:
@pfalcon I think we should untar the micropython-upip-*.tar.gz file to the build/ directory.  It's generated content after all, and will get cleaned via make clean if it's in build.

pfalcon commented:
> @pfalcon I think we should untar the micropython-upip-*.tar.gz file
> to the build/ directory.  It's generated content after all, and will
> get cleaned via make clean if it's in build.

Done in a546acda8c4c1342ee5f0e50733915e15a775756.

dpgeorge commented:
Thanks!  Also what do you think about frozen_upip?  Shouldn't that also go in build/?

pfalcon commented:
> Thanks! Also what do you think about frozen_upip? Shouldn't that also go in build/?

Yes, done in c36635c112eee93dd987b98ca668fe488470aed0

dpgeorge commented:
Great, nice and tidy now!
